server/core/runtime/multienv_step_runner.go

```diff
@@ -0,0 +1,53 @@
+package runtime
+
+import (
+	"encoding/json"
+
+	"strings"
+
+	"github.com/runatlantis/atlantis/server/events/models"
+)
+
+type KeyValue struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+type MultiEnvCallerResult struct {
+	Success      bool
+	ErrorMessage string `json:"errorMessage"`
+	Result       []KeyValue
+}
+
+// EnvStepRunner set environment variables.
+type MultiEnvStepRunner struct {
+	RunStepRunner *RunStepRunner
+}
+
+// Run runs the multienv step command.
+// The command must return a json string containing the array of name-value pairs that are being added as extra environment variables
+func (r *MultiEnvStepRunner) Run(ctx models.ProjectCommandContext, command string, path string, envs map[string]string) (string, error) {
+	res, err := r.RunStepRunner.Run(ctx, command, path, envs)
+	if err == nil {
+		var callerResult MultiEnvCallerResult
+		err = json.Unmarshal([]byte(res), &callerResult)
+		if err == nil {
+			if callerResult.Success {
+				if len(callerResult.Result) > 0 {
+					var sb strings.Builder
+					sb.WriteString("Dynamic environment variables added:\n")
+					for _, item := range callerResult.Result {
+						envs[item.Name] = item.Value
+						sb.WriteString(item.Name)
```

Comment:
Please let me share why we need so detailed logging.
Within our company we have a dedicated team that technically owns Atlantis (and other stuffs, but this is not relevant), only this team has the knowledge how Atlantis works, how the architecture looks like. There are dozens of other teams whose Terraform developer members only know that commenting to their Azure DevOps PRs plans and applies Terraform related infra. They only aware of their own repo and a connected Azure subscription. An automated self service ticketing mechanism is provided for them to select a repo and assign to a Azure subscription (with proper permission handling), so they does not even know how their own repo is connected to their subscription, they only focus on writing Terraform code in order to set up their own Azure infra.
Let's see our current use case. Somebody needs a team specific secret in his/her Terraform code e.g. a specific Terraform provider requires it in the form of an environment variable (this is a very common scenario). We (the Atlantis owner hub team) do not want to add project specific envvars to our company specific Atlantis workflow (containing custom checks by 3rd party tools integrated), and the members of the specific project does not have the permission to create their own workflow (to avoid overriding the company standards/policy).
Instead, this the team member can pass the secret via a specific file in the repo that references a key vault hosting the secrets. He does not know anything but he needs to have a committed file in his repo, the file contains the name of a key vault and the access permissions need to be set on the key vault properly. Atlantis workflow extracts the secrets from the key vault via a web api call and creates the required environment variables under the hood.
The webservice must perform several actions where multiple errors can occur, e.g.:
- The committed key vault file contains invalid reference
- The key vault does not exist
- The team specific Azure deployment user does not have the proper permission to access the key vault
- Missing permission to read the secrets from the key vault
- Atlantis workflow is not able to call the web api, maybe network issue
These notifications must be passed to the UI (which is the Azure DevOps PR in our case) to let the users resolve the issues.
Sometimes the key vault reference is correct and the permissions to access the key vault are also correct, Atlantis workflow successfully creates the environment variables, but Terraform does not find the specific envvar. Maybe the user unintentionally referenced an other key vault or meanwhile somebody else removed the required secrets from the vault.
That is why it's essential to display the dynamic envvar names through Atlantis PR comment, and let the user view what kind of dynamic envvars are provided to Terraform.
Anyway, according to our security review displaying the name (only the name, not he value!) of the dynamic envvars is not a big security risk compared what we can win, our goal is providing a self service Azure infra building process, avoiding the tickets containing "why the hell Terraform cannot see the envvars" that probaly would lead to manual intervention one by one wasting our resources.